File: res2pat.py
import math
import sys

semvector = [[1,0,0],[0,1,0],[0,0,1]]

# initialize array with a certain numer of indices
def make_array(indices, init):
	return [init for x in xrange(indices)]

# Create the char arrays with 90 slots
# [title, date, undef, pats, ins, outs] = [make_array(90,0) for x in xrange(6)]

# Create the char arrays with 120 slots
# [line, dit] = [make_array(120) for x in xrange(2)]

# Create the rest of the variables, though they can be implicitly defined

# Create the float array
raw = make_array(80,0.0)

# Call the input file f1

def main():
	# declare a string p
	# and three ints i,j,k

	# open the input file
	try: 
		with open(sys.argv[1], "r") as f1:
			title = f1.read(81)
			date = f1.read(81)
			undef = f1.read(81)
			pats = f1.read(81)
			ins = f1.read(81)
			outs = f1.read(81)
			undef = f1.read(81)
			undef = f1.read(81)
			undef = f1.read(81)
		 	
		 	print("SNNS pattern definition file V1.4\n")
		 	print("%s\n\n" % date)
		 	print("%s" % pats)
		 	print("%s" % ins)
		 	print("%s" % outs)
		 
		 	pat = findint(pats)
			In = findint(ins)
			out = findint(outs)

			for i in range(0,pat):
				# comment line
				line = f1.read(81)
				print("%s" % line)

				# input line 
				for j in range(0, In):
					# i know this is wrong
					dit = string(f1)
					print("%s", dit)
					# I'm confused about this code. It needs to mimmick:
					#fscanf(f1, "%s", dit);
					#printf("%s ", dit);

				# output line
				for k in range(0, out):
					# what does this ampersand mean?
					# this needs to mimmick
					#fscanf(f1, "%f", &raw[k]);
					raw[k] = float(f1)
					# i know this is wrong
				distf()
				print("\n")
				line = f1.read(10)
		except IOError:
			print ("Cannot open file")
			sys.exit(1)


# Compute the Euclidean calculation of error
def distf():
	bestdist = 100000.0
	bestpointer = 0
	for j in range(0,3):
		dist = 0.0
		for k in range(0,3):
			temp = semvector[j][k] - raw[k]
			dist += temp * temp
		if dist < bestdist:
			bestdist = dist;
			bestpointer = j;

	#print it out
	for k in range(0,3):
		print("%s ", % semvector[bestpointer][k])

# Get the  number which is in a char line
def findint(text): #text is a pointer to a character
	i = 0
	j = 0
	while text[i]:
		if (text[i].isdigit()):
			text[j] = text[i]
			j += 1
		i += 1
	text[j] = ""
	return(int(''.join(text))

File: Polinsky_VanEverbroeck_2003/res2pat.c
#include <stdio.h>
#include <string.h>
#include <math.h>
#include <stdlib.h>

float semvector [3] [3] = {
{1, 0, 0},
{0, 1, 0},
{0, 0, 1},
}; 

int distf(void);
int findint(char *);

char title[90], date[90], undef[90], pats[90], ins[90], outs[90];
char line[120], dit[120];
int pat, in, out, i, j, k, l;
float dist, bestdist, temp;
int bestpointer;
float raw[80];
FILE *f1;

main (int argc, char *argv[])  {

char *p;
int i, j, k;

if ((f1 = fopen(argv[1], "r" )) == NULL) {
	printf("\nERROR: Cannot open file.\n");
	exit(0);
}

fgets(title, 81, f1);
fgets(date, 81, f1);
fgets(undef, 81, f1);
fgets(pats, 81, f1);
fgets(ins, 81, f1);
fgets(outs, 81, f1);
fgets(undef, 81, f1);
fgets(undef, 81, f1);
fgets(undef, 81, f1);

printf("SNNS pattern definition file V1.4\n");
printf("%s\n\n", date);
printf("%s", pats);
printf("%s", ins);
printf("%s\n", outs);

pat = findint(pats);
in = findint(ins);
out = findint(outs);

for (i=0 ; i<pat; i++) {

	// comment line
	fgets(line, 81, f1);
	printf("%s", line);
	
	// input lines
	for (j=0 ; j<in; j++) {
		fscanf(f1, "%s", dit);
		printf("%s ", dit);
	}
	printf("\n");
	
	// output lines
	for (k=0; k<out; k++) {
		// for (l=0; l<3; l++) {
			fscanf(f1, "%f", &raw[k]);
		// }
	}
	distf();
	printf("\n");
	fgets(line, 10, f1);
}

} // end main()	

/* This section computes Euclidean calculation of error */
int distf(void) {
int i,j,k;
bestdist = 100000.0;
bestpointer = 0;

for (j=0; j<3; j++) {
	dist = 0.0;

	for (k=0; k<3; k++) {
	      temp = semvector[j][k] - raw[k];
	      dist += temp * temp ;
	}

	// dist = sqrt(dist);       
	if (dist < bestdist) {
		// printf("---> %d : %f \n", j, dist);
		bestdist = dist;
		bestpointer = j;
	}
}
	
// now print it out
for ( k=0; k<3; k++) {
	// printf("%3.1f ", semvector[bestpointer][k]);
	printf("%d ", (int) semvector[bestpointer][k]);
}

} // end distf


/* get the number which is in a char line */
int findint(char * text) {
int i = 0;
int j = 0;

while (text[i] != '\0') {
	if (isdigit(text[i])) {
		text[j] = text[i];
		j++;
	} else {
	}
	i++;
}

text[j] = '\0';
i = atoi(text);

return(i);

} // end findint()
